src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.cs

```diff
@@ -124,14 +124,23 @@ public static async Task<ServiceHubRemoteHostClient> CreateWorkerAsync(Workspace
 
         private ServiceHubRemoteHostClient(
             Workspace workspace,
+            TraceSource logger,
             ConnectionManager connectionManager,
             Stream stream)
             : base(workspace)
         {
             _connectionManager = connectionManager;
 
-            _rpc = new JsonRpc(new JsonRpcMessageHandler(stream, stream), target: this);
-            _rpc.JsonSerializer.Converters.Add(AggregateJsonConverter.Instance);
+            var jsonFormatter = new JsonMessageFormatter()
+            {
+                JsonSerializer = { Converters = { AggregateJsonConverter.Instance } }
+            };
+
+            _rpc = new JsonRpc(new HeaderDelimitedMessageHandler(stream, jsonFormatter), target: this)
+            {
+                CancelLocallyInvokedMethodsWhenConnectionIsClosed = true,
+                TraceSource = logger
+            };
```

Comment:
Pattern seems to repeat a bit. Worth putting in a helper?

Reply:
yep. did that places I can. between VS and Servicehub there is no layer that are shared with JsonRpc dependency so, those are still duplicated.

Reply:
Sure, but we can have linked files right?

Reply:
sure. used the linked file to share code.